h2/src/main/org/h2/mvstore/tx/TransactionMap.java

```diff
@@ -622,16 +622,25 @@ protected TMIterator(TransactionMap<K,?> transactionMap, K from, K to, boolean i
             this.transactionMap = transactionMap;
             TransactionStore store = transactionMap.transaction.store;
             MVMap<K, VersionedValue> map = transactionMap.map;
+            // The purpose of the following loop is to get a coherent picture
+            // of a state of two independent volatile / atomic variables
+            // which they had at some recent moment in time.
+            // In order to get such a "snapshot", we wait for a moment of silence
+            // when neither of the variables concurrently chenges it's value.
             BitSet committingTransactions;
             MVMap.RootReference mapRootReference;
             do {
                 committingTransactions = store.committingTransactions.get();
                 mapRootReference = map.getRoot();
             } while (committingTransactions != store.committingTransactions.get());
-
+            // Now we have a snapshot, where mapRootReference points to state of the map
```

Comment:
OK, such comments make sense.

I don't see way to use the shared code in all such places, but perhaps these loops can at least have the same implementation everywhere, now they are different (and they were different before your changes). Only the `committingTransactions` is checked above, but in the code below `mapRootReference` is also compared:
https://github.com/h2database/h2database/blob/0c46e47e7824f4fd366f0d7cebd91f6e45f106ed/h2/src/main/org/h2/mvstore/tx/TransactionMap.java#L103-L104
It this second comparison needed? If it is, why there is no such comparison in `TMIterator()`?

Reply:
Oh, this comparison is not needed here and is needed in `sizeAsLong()`. Both loops read some values and check that all values with exception of the last one are still the same. This logic is not oblivious.

I do not have complains any more.